libtcmu.c

```diff
@@ -618,7 +797,19 @@ static void remove_device(struct tcmulib_context *ctx,
 
 	darray_remove(ctx->devices, i);
 
+        struct tcmu_aio_queue *aio_wq = &dev->work_queue;
+
+        /*
+         * The order of cleaning up worker threads and calling ->removed()
+         * is important: for sync handlers, the worker thread needs to be
+         * terminated before removing the handler (i.e., calling handlers
+         * ->close() callout) in order to ensure that no store callouts
+         * are getting invoked when shutting down the handler.
+         */
+        cancel_thread(aio_wq->aio_wq_thread);
         dev->handler->removed(dev);
```

Comment:
In case of librbd, the outstanding I/O request should get completed (upon call to `rbd_close()`). That would cleanup the tracking queue (via call from request aio completion). Other handlers should also properly cleanup in-flight I/Os when shutdown.